### ISSUE TITLE
Is this actually "encoder-decoder" vs. standard many-to-many?

### ISSUE THREAD

**Issue by wasd12345** [Arturus/kaggle-web-traffic#22]
Thanks for sharing your code, very helpful.

From your computational graph and model code, it looks like the "decoder" at each timestep takes 2 inputs:
1) the previous hidden state from the decoder (hidden state of deocder GRU cell at previous timestep)
2) a concatenated vector of inputs = [previous prediction, features, attention] where attention is optional. 

The first timestep decoder cell gets the "encoded state" as the last hidden state of the encoder. But future decoder timesteps do NOT get this encoded representation again. So the computational graph does not look like that in the original RNN encoder-deocder paper or like in the seq2seq encoder-decoder section of the Deep Learning book. I.e. it seems this model architecture is more like a standard many-to-many RNN but not encoder-decoder, right? I.e. you do not feed in the encoded state "c" again?

Thanks

Cho Encoder-decoder:
https://arxiv.org/pdf/1406.1078.pdf
Fig 1. on pg. 2
<img width="311" alt="cho encoder-decoder" src="https://user-images.githubusercontent.com/13309204/42912023-2f86fcf4-8aa2-11e8-8c89-a35fbc6629cf.png">

Deep Learning Book:
http://www.deeplearningbook.org/contents/rnn.html
Section 10.4 Encoder-Decoder Sequence-to-Sequence Architectures
pg. 391 Fig 10.12
<img width="480" alt="encoder decoder dl book" src="https://user-images.githubusercontent.com/13309204/42912098-8246c2c6-8aa2-11e8-8ba1-76f0ddf6a5d6.png">

Your model:
https://github.com/Arturus/kaggle-web-traffic/blob/master/how_it_works.md#model-core
<img width="895" alt="kaggle model" src="https://user-images.githubusercontent.com/13309204/42912104-891939b2-8aa2-11e8-86b7-d0e1f7ebb6f9.png">

**Comment by wasd12345** [comment]
See this fork for the same model with the more typical encoder-decoder option:
[https://github.com/wasd12345/kaggle-web-traffic](url)

--Technically, the above linked fork is not quite like the deep learning book either. It does not provide the context vector **C** when predicting **y**, instead it only provides context to the decoder states but at **every** timestep, not just the first decoder timestep.

After very basic comparison [just run both modes on my data for 10 epochs] of with vs. without providing the encoded state context vector C, seems like it does help: notice lower SMAPE for the encoder-decoder option:

NO encoder-decoder (as is):
<img width="1122" alt="no encoder decoder" src="https://user-images.githubusercontent.com/13309204/42963199-64cbbf70-8b48-11e8-8963-407651b7b141.png">

vs.

WITH more typical encoder-decoder (in the linked fork):
<img width="1115" alt="with encoder decoder" src="https://user-images.githubusercontent.com/13309204/42963227-7778be02-8b48-11e8-82e4-fd12c19c59f7.png">

Walk-forward validation SMAPEs printed out are consistently 1% to 2% better now. [**Also, these printed validation SMAPEs are the ones printed out during training phase, which is using an epsilon smoothed SMAPE on log1p transformed data. So the actual numbers are irrelevant but the point is the encoder-decoder context is better on a relative basis]


The wasd12345 fork uses a computational graph like this:
<img width="1043" alt="my comp graph" src="https://user-images.githubusercontent.com/13309204/42963557-68b05be0-8b49-11e8-8338-7a1e3568461e.png">

**Comment by yuhaomin** [comment]
How can I show the value of the loss?

**Comment by yuhaomin** [comment]
Thank you very much for your answer. But my results can't show smape loss, mae loss,... The result  just shows Nan value.[ Best top SMAPE=nan  (), frwd/side best MAE=nan/nan, SMAPE=nan/nan; avg MAE=nan/nan, SMAPE=nan/nan, 3 active models].  I'm not sure how to display the values.

**Comment by wasd12345** [comment]
@yuhaomin   in his readme, the original branch Arturus gave the training command like this:
`python trainer.py --name s32 --hparam_set=s32 --n_models=3 --name s32 --no_eval --no_forward_split --asgd_decay=0.99 --max_steps=11500 --save_from_step=10500`

That command is not doing evaluation, so:
Get rid of the no_eval option to do evaluation. 
Remove the no_forward_split to do forward split evaluation, so use something like this:
`python trainer.py --name=s32 --hparam_set=s32 --n_models=3 --asgd_decay=0.99 --max_steps=11500 --save_from_step=1000 --max_epoch=50`

Are you using your own data? If yes, make sure you don't have Nans in the data that corrupt the encoder state.

**Comment by wasd12345** [comment]
No, I don't think I had that error, but I did have other issues with all Nan encoder state leading to all NAN metrics that were related to the train_completeness_threshold.
You're using the same python/tensorflow versions? Are you using this Arturus branch or the wasd12345 fork? You're using that Kaggle data or your own?

**Comment by yuhaomin** [comment]
Thank you!  May be the error caused by version of tensorflow (tensosflow 1.6.)  And I also try to use my data use by this project.

**Comment by Arturus** [comment]
@wasd12345 I did not think that providing last encoder state for **every** decoder step would help, but as your results show, it really helps to get better results.
So yes, your "classic" decoder variant is better than mine, thank you for sharing this!

Btw, I updated model code to work with TF 1.10
